### ISSUE TITLE
Dynamic slots error with generics

### ISSUE THREAD

**Issue by csisy** [vuejs/language-tools#4538]
### Vue - Official extension or vue-tsc version

2.0.26

### VSCode version

1.91.0

### Vue version

3.4.31

### TypeScript version

5.5.3

### Steps to reproduce

Trying to build a table component which has a single generic parameter, representing the type of the data row. Column definitions are given according to this type.

```typescript
<script setup lang="ts" generic="T extends { [K in keyof T]: T[K] } & { id: PropertyKey }">

// definition of a column
export type ColumnDef<T> = {
  // the property is a key of the generic type
  property: keyof T & string;
  title: string;
};
```
I'd like to define the slot names and the corresponding props:
```typescript
// this is the type of the column template; receives the given row and the value based on the given 'property' of the column def.
type ColumnTemplateData<K extends keyof T> = {
  row: T;
  value: T[K];
};

// we will have slots in the `column-{keyof T}` format with these props
defineSlots<{
  [K in keyof T as `column-${K & string}`]: (_: ColumnTemplateData<K>) => any;
}>();
```
Try to define a slot with a generated name:
```html
<div v-for="col in columns" :key="col.property">
  <!-- error: Type '`column-${UnwrapRef<keyof T & string>}`' cannot be used to index type [...] -->
  <slot :name="`column-${col.property}`" :row="row" :value="row[col.property]">
    <span>{{ row[col.property] }}</span>
  </slot>
</div>
```

### What is expected?

Based on other threads and PR [#8374](https://github.com/vuejs/core/pull/8374), this should report no errors.

### What is actually happening?

When the slot is defined an error is generated:
```
Type '`column-${UnwrapRef<keyof T & string>}`' cannot be used to index type 'Readonly<{ [K in keyof T as `column-${K & string}`]: (_: ColumnTemplateData<K>) => any; }> & { [K in keyof T as `column-${K & string}`]: (_: ColumnTemplateData<K>) => any; }'.ts-plugin(2536)
```

### Link to minimal reproduction

https://github.com/csisy/vue-dynamic-slots-repro

### Any additional comments?

If you try the example (`npm run dev`), you can see that it actually works runtime.

**Comment by csisy** [comment]
Might be related? #3787

**Comment by KermanX** [comment]
I think this is a TypeScript limitation. See [this code](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABMAPAFUQUwB5U2AEwGdEBvRAbQGlEYkBrTATzmETQF0Aud6jxAL6IAZGVoEeABQBOcAA6ZpUJlWaCAfAAoAlGQBQiQ4mULEAYTgAbEAFswAEUyo06xAF59Rr4jmyFSph5GFjYMUSIoaToAcwBuAyMBeITDCAQIxCJLOCgiHgAlTABDAgRLJhRyaloGZlZ2RCKSAAM06zsAWgASUhpwyJiBZu5ETQB9HipdN1cisCZYjXdEMBBLS0aSOYWUxDSwDLbbMB4LdocndFcPXa9yX3lFZR4AchgCF4AaYxgoS0xXgBJewvQSfXYcZJeLI5IgUVpWY7dUhHOxwgAMHAAdA9-Mohhw9AI9EA) - `f1` errors but `f2` is ok.
